daemon/cmd/local_node_init.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	k8sLabels "k8s.io/apimachinery/pkg/labels"
+
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/node/addressing"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type localNodeInitializerParams struct {
+	cell.In
+
+	Config    *option.DaemonConfig
+	Clientset client.Clientset
+	LocalNode k8s.LocalNodeResource
+}
+
+// localNodeInitializer performs the bootstrapping of the LocalNodeStore,
+// which contains information about the local Cilium node populated from
+// configuration and Kubernetes.
+type localNodeInitializer struct {
+	localNodeInitializerParams
+}
+
+func (ini *localNodeInitializer) InitLocalNode(ctx context.Context, n *node.LocalNode) error {
+	n.Source = source.Local
+
+	if err := ini.initFromConfig(ctx, n); err != nil {
+		return err
+	}
+
+	if err := ini.initFromK8s(ctx, n); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newLocalNodeInitializer(p localNodeInitializerParams) node.LocalNodeInitializer {
+	return &localNodeInitializer{p}
+}
+
+func (ini *localNodeInitializer) initFromConfig(ctx context.Context, n *node.LocalNode) error {
+	// If there is one device specified, use it to derive better default
+	// allocation prefixes
+	node.SetDefaultPrefix(ini.Config, ini.Config.DirectRoutingDevice, n)
+
+	// Initialize node IP addresses from configuration.
+	if ini.Config.IPv6NodeAddr != "auto" {
+		if ip := net.ParseIP(ini.Config.IPv6NodeAddr); ip == nil {
+			return fmt.Errorf("invalid IPv6 node address: %q", ini.Config.IPv6NodeAddr)
+		} else {
+			if !ip.IsGlobalUnicast() {
+				return fmt.Errorf("Invalid IPv6 node address: %q not a global unicast address", ip)
+			}
+			n.SetNodeInternalIP(ip)
```

Comment:
no-op: I note that the previous version did `node.SetIPv[46]()` and the new one directly sets the internalIP. However, the per-protocol functions just did exactly this anyway. It's a subtle change to find in a big commit, but it seems like it's not functionally changing the behaviour.

Reply:
nit: I glossed over other similar changes where previous code did IPv4-specific or IPv6-specific logic and the new ones use the generic functions, since they seem to be iterating over the slices from the k8s objects instead of doing hardcoded family-specific calls. Hopefully our code coverage is good enough to pick up if there are any unintended behaviour changes related to that. Otherwise if we want to validate this with code review, I'd suggest breaking the commit down further to make it trivially clear that there's no change in behaviour.

Reply:
The old `SetIPv[46]` went through the same code path anyway, so there's no functional change here.